testbed/testbed/utils/libp2pHTTP.go

```diff
@@ -0,0 +1,139 @@
+package utils
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	files "github.com/ipfs/go-ipfs-files"
+	ipld "github.com/ipfs/go-ipld-format"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	gostream "github.com/libp2p/go-libp2p-gostream"
+	p2phttp "github.com/libp2p/go-libp2p-http"
+	mh "github.com/multiformats/go-multihash"
+)
+
+type Libp2pHTTPNode struct {
+	client *http.Client
+	h      host.Host
+	svr    *http.Server
+}
+
+func CreateLibp2pHTTPNode(ctx context.Context, h host.Host, nodeTP NodeType) (*Libp2pHTTPNode, error) {
+	switch nodeTP {
+	case Seed:
+		// Server
+		listener, err := gostream.Listen(h, p2phttp.DefaultP2PProtocol)
+		if err != nil {
+			return nil, err
+		}
+		// start an http server on port 8080
+		svr := &http.Server{}
+		go svr.Serve(listener)
+		time.Sleep(1 * time.Second)
+		return &Libp2pHTTPNode{
+			h:   h,
+			svr: svr,
+		}, nil
+	case Leech:
+		tr := &http.Transport{}
+		tr.RegisterProtocol("libp2p", p2phttp.NewTransport(h))
+		client := &http.Client{Transport: tr}
+
+		return &Libp2pHTTPNode{
+			client: client,
+			h:      h,
+		}, nil
+	default:
+		return nil, errors.New("nodeType NOT supported")
+	}
+}
+
+func (l *Libp2pHTTPNode) Add(ctx context.Context, file files.Node) (cid.Cid, error) {
+	f := files.ToFile(file)
+	if f == nil {
+		return cid.Undef, errors.New("node is NOT a File")
+	}
+
+	c, err := randCid()
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	// set up http server to send file
+	http.HandleFunc(fmt.Sprintf("/%s", c.String()), func(w http.ResponseWriter, r *http.Request) {
+		defer f.Close()
+		_, err := io.Copy(w, f)
+		if err != nil {
+			panic(err)
+		}
+	})
+
+	return c, nil
+}
+
+func (l *Libp2pHTTPNode) Fetch(ctx context.Context, cid cid.Cid, peers []PeerInfo) (files.Node, error) {
+	seedCount := 0
+	var seed peer.ID
+
+	for _, p := range peers {
+		if p.Nodetp == Seed {
+			seedCount++
+			seed = p.Addr.ID
+		}
+	}
+	if seedCount != 1 {
+		return nil, errors.New("libp2p http should ONLY have one seed")
+	}
+
+	resp, err := l.client.Get(fmt.Sprintf("libp2p://%s/%s", seed.String(), cid.String()))
+	if err != nil {
+		return nil, err
+	}
+
+	return files.NewReaderFile(resp.Body), nil
+}
+
+func (l *Libp2pHTTPNode) Host() host.Host {
+	return l.h
+}
+
+func (l *Libp2pHTTPNode) ClearDatastore(ctx context.Context, rootCid cid.Cid) error {
+
+	return nil
+}
+
+// NOOP FOR now,
+// #TODO Fix
+func (l *Libp2pHTTPNode) EmitMetrics(recorder MetricsRecorder) error {
+	return nil
+}
+
+// NO-OP
+func (l *Libp2pHTTPNode) DAGService() ipld.DAGService {
+	return nil
+}
+
+// NO-OP
+func (l *Libp2pHTTPNode) EmitKeepAlive(recorder MessageRecorder) error {
+	return nil
+}
+
+func randCid() (cid.Cid, error) {
```

Comment:
Out of a curiosity, why are you using a random CID instead of using the CID for the file? 🙈

Reply:
Since there's no import to a blockstore, there's no IPLD graph generated, and therefor no obvious CID generated.

However, I think the right thing to do here, and it's a relatively small change, is to multihash sum all the bytes in the file, instead of a random integer. That will mean that the CID is in fact the correct CID for a raw block that contains all the bytes of the file, and more importantly will make it deterministic based on the input of the file.

The downside of random CID is if at some point we want to use more than one seed, we're going to end up with each seed generating different CIDs. The suggested approach of summing all the bytes in the file should enable all seeds to generate the same CID.

Reply:
The problem here is that I'd have to read the file to get it's hash. But, we need to read it again during sending it to the user.  I tried using the `Seek` method on the `files.File` interface to reset the offset after hashing it but looks like it isn't always supported.

Is there a clean way of reading this same `files.File` twice ?

Reply:
You could certainly use an io.TeeReader to write into a secondary temporary file as you hash and serve that.

Reply:
non-blockign for this PR I think -- just something you're going to need to do to support more than one seed.